File: App/Resources/AlgorithmDetails/medianmergesort/medianmergesort.cpp
#include <cstdio>
#include <utility>
#include <vector>

int array[24] = {0,  39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81,
                 68, 83, 32, 56, 10, 2,  95, 46, 21, 74, 6,  38};

void printList(int items[], int size) {
  printf("[");
  if (size > 0) {
    printf("%d", items[0]);
    for (int i = 1; i < size; i++) {
      printf(", %d", items[i]);
    }
  }
  printf("]");
}

void mergeSort(int arr[], std::vector<int> &scratch, int start, int end) {
  if (end - start < 2)
    return;
  int middle = (start + end) / 2;
  mergeSort(arr, scratch, start, middle);
  mergeSort(arr, scratch, middle, end);
  int left = start, right = middle, dest = start;
  while (left < middle && right < end) {
    scratch[dest++] = arr[left] <= arr[right] ? arr[left++] : arr[right++];
  }
  while (left < middle)
    scratch[dest++] = arr[left++];
  while (right < end)
    scratch[dest++] = arr[right++];
  for (int i = start; i < end; i++)
    arr[i] = scratch[i];
}

int medianOfThree(int x, int y, int z) {
  if (x > y)
    std::swap(x, y);
  if (y > z)
    std::swap(y, z);
  if (x > y)
    std::swap(x, y);
  return y;
}

void sort(int arr[], int n) {
  std::vector<int> scratch(n);
  int start = 0, end = n;
  while (end - start > 16) {
    int pivot =
        medianOfThree(arr[start], arr[(start + end - 1) / 2], arr[end - 1]);
    int left = start, right = end - 1;
    while (left <= right) {
      while (left <= right && arr[left] < pivot)
        left++;
      while (left <= right && arr[right] > pivot)
        right--;
      if (left <= right)
        std::swap(arr[left++], arr[right--]);
    }
    if (left == start || left == end) {
      mergeSort(arr, scratch, start, end);
      return;
    }
    if (left - start <= end - left) {
      mergeSort(arr, scratch, start, left);
      start = left;
    } else {
      mergeSort(arr, scratch, left, end);
      end = left;
    }
  }
  for (int i = start + 1; i < end; i++) {
    int value = arr[i], j = i;
    while (j > start && arr[j - 1] > value) {
      arr[j] = arr[j - 1];
      j--;
    }
    arr[j] = value;
  }
}

int main(int argc, char *argv[]) {
  int size = sizeof(array) / sizeof(array[0]);
  sort(array, size);
  printList(array, size);
  return 0;
}

File: App/Resources/AlgorithmDetails/medianmergesort/medianmergesort.rb
def sort(arr)
  scratch = Array.new(arr.length, 0)
  merge_sort = lambda do |start, finish|
    if finish - start > 1
      middle = (start + finish) / 2
      merge_sort.call(start, middle)
      merge_sort.call(middle, finish)
      left = start
      right = middle
      dest = start
      while left < middle && right < finish
        if arr[left] <= arr[right]
          scratch[dest] = arr[left]
          left += 1
        else
          scratch[dest] = arr[right]
          right += 1
        end
        dest += 1
      end
      while left < middle
        scratch[dest] = arr[left]
        left += 1
        dest += 1
      end
      while right < finish
        scratch[dest] = arr[right]
        right += 1
        dest += 1
      end
      (start...finish).each { |i| arr[i] = scratch[i] }
    end
  end

  start = 0
  finish = arr.length
  while finish - start > 16
    pivot = [arr[start], arr[(start + finish - 1) / 2], arr[finish - 1]].sort[1]
    left = start
    right = finish - 1
    while left <= right
      left += 1 while left <= right && arr[left] < pivot
      right -= 1 while left <= right && arr[right] > pivot
      if left <= right
        arr[left], arr[right] = arr[right], arr[left]
        left += 1
        right -= 1
      end
    end
    if left == start || left == finish
      merge_sort.call(start, finish)
      return
    end
    if left - start <= finish - left
      merge_sort.call(start, left)
      start = left
    else
      merge_sort.call(left, finish)
      finish = left
    end
  end

  (start + 1...finish).each do |i|
    value = arr[i]
    j = i
    while j > start && arr[j - 1] > value
      arr[j] = arr[j - 1]
      j -= 1
    end
    arr[j] = value
  end
end

array = [0, 39, 21, 62, 91, 77, 14, 23,
  90, 69, 51, 81, 68, 83, 32, 56,
  10, 2, 95, 46, 21, 74, 6, 38]
sort(array)
p array
